### PR TITLE
CLOUDP-57522: Creating Atlas Alert Configuration

### PR DESCRIPTION
**Pull request by andreaangiolillo**
## Proposed changes

_Jira ticket:_ [CLOUDP-57522](https://jira.mongodb.org/browse/CLOUDP-57522/[name-of-ticket])

## Checklist

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run `make fmt` and formatted my code

### CODE REVIEW

File: e2e/atlas_alertconfig_test.go [hunk 1]
```diff
@@ -36,6 +47,61 @@ func TestAtlasAlertConfig(t *testing.T) {
 	atlasEntity := "atlas"
 	alertConfigEntity := "alert-config"
 
+	t.Run("Create", func(t *testing.T) {
```
Comment (gssbzn) [inline e2e/atlas_alertconfig_test.go]: Let's not do e2e until we can delete it or will be creating a bunch of useless alerts

File: internal/cli/atlas_alertconfig_create.go [hunk 2]
```diff
@@ -0,0 +1,221 @@
+// Copyright 2020 MongoDB Inc
```
Comment (gssbzn) [inline internal/cli/atlas_alertconfig_create.go]: [nit] naming should be `atlas_alert_config_create.go`

File: internal/cli/atlas_alertconfig_create.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+	notificationChannelName         string
+	apiKey                          string // notificationsDatadogApiKey, notificationsOpsGenieApiKey, notificationsVictorOpsApiKey
+	notificationEmailAddress        string
+	notificationFlowName            string
+	notificationMobileNumber        string
+	notificationRegion              string // notificationsOpsGenieRegion, notificationsDatadogRegion
+	notificationOrgName             string
+	notificationServiceKey          string
+	notificationTeamID              string
+	notificationTypeName            string
```
Comment (gssbzn) [inline internal/cli/atlas_alertconfig_create.go]: [nit] could this be just `notificationType` the name part is a bit redundant

File: internal/cli/atlas_alertconfig_create.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+	*globalOpts
+	event                           string
+	matcherFieldName                string
+	matcherOperator                 string
+	matcherValue                    string
+	metricThresholdMetricName       string
+	metricThresholdOperator         string
+	metricThresholdUnits            string
+	metricThresholdMode             string
+	token                           string // notificationsApiToken, notificationsFlowdockApiToken
```
Comment (gssbzn) [inline internal/cli/atlas_alertconfig_create.go]: [q] should this be `notificationToken` to follow your naming structure?

File: internal/usage/usage.go [hunk 5]
```diff
@@ -34,14 +34,39 @@ const (
...
-	WhitelistType = `Type of whitelist entry.
-Valid values: cidrBlock|ipAddress`
-	Service = `Type of MongoDB service.
-Valid values: cloud|cloud-manager|ops-manager`
-	Provider = `Name of your cloud service provider.
-Valid values: AWS|AZURE|GCP.`
+	WhitelistType = `Type of whitelist entry. Valid values: cidrBlock|ipAddress`
+	Service       = `Type of MongoDB service. Valid values: cloud|cloud-manager|ops-manager`
+	Provider      = `Name of your cloud service provider.
+										Valid values: AWS|AZURE|GCP.`
```
Comment (gssbzn) [inline internal/usage/usage.go]: [q] what's with the spacing here? unintended?

File: internal/store/alert_configuration.go [hunk 6]
```diff
@@ -40,3 +45,16 @@ func (s *Store) AlertConfigurations(projectID string, opts *atlas.ListOptions) (
 		return nil, fmt.Errorf("unsupported service: %s", s.service)
 	}
 }
+
+// CreateAlertConfiguration encapsulate the logic to manage different cloud providers
+func (s *Store) CreateAlertConfiguration(alert *atlas.AlertConfiguration) (*atlas.AlertConfiguration, error) {
```
Comment (gssbzn) [inline internal/store/alert_configuration.go]: [nit] instead of alert let's do something like, `ac`(alert config, go likes short names), or `config` or `alertConfig`, will add alerts later so trying to to confuse this with that

File: internal/store/alert_configuration.go [hunk 7]
```diff
@@ -40,3 +45,16 @@ func (s *Store) AlertConfigurations(projectID string, opts *atlas.ListOptions) (
...
+	switch s.service {
+	case config.CloudService:
+		project := alert.GroupID
+		alert.GroupID = ""
```
Comment (gssbzn) [inline internal/store/alert_configuration.go]: can you add a `// TODO: fix me" and link to the PR where we fix this on the go client?

File: internal/cli/atlas_alertconfig_create.go [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+	case flowdock:
+		notification.FlowdockAPIToken = opts.token
+		notification.FlowName = opts.notificationFlowName
+		notification.OrgName = opts.notificationOrgName
+
+	case sms:
+		notification.MobileNumber = opts.notificationMobileNumber
+
+	case group, user, org:
+		notification.SMSEnabled = &opts.notificationSmsEnabled
```
Comment (gssbzn) [inline internal/cli/atlas_alertconfig_create.go]: all your uses of `opts.notificationSmsEnabled` is with `&` you could define it as a pointer in the struct and not use the `&` part

File: internal/cli/atlas_alertconfig_create.go [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+		notification.EmailEnabled = &opts.notificationEmailEnabled
```
Comment (gssbzn) [inline internal/cli/atlas_alertconfig_create.go]: Same comment as with sms

File: internal/cli/atlas_alertconfig_create.go [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+	alert.GroupID = opts.ProjectID()
+	alert.EventTypeName = strings.ToUpper(opts.event)
+	alert.Enabled = &opts.enabled
+
+	if opts.matcherFieldName != "" {
+		match := new(atlas.Matcher)
+		match.FieldName = strings.ToUpper(opts.matcherFieldName)
+		match.Operator = strings.ToUpper(opts.matcherOperator)
+		match.Value = strings.ToUpper(opts.matcherValue)
+		alert.Matchers = []atlas.Matcher{*match}
```
Comment (gssbzn) [inline internal/cli/atlas_alertconfig_create.go]: can we encapsulate this in a function?

File: internal/cli/atlas_alertconfig_create.go [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+	}
+
+	if opts.metricThresholdMetricName != "" {
+		metric := new(atlas.MetricThreshold)
+		metric.MetricName = strings.ToUpper(opts.metricThresholdMetricName)
+		metric.Operator = strings.ToUpper(opts.metricThresholdOperator)
+		metric.Threshold = opts.metricThresholdThreshold
+		metric.Units = strings.ToUpper(opts.metricThresholdUnits)
+		metric.Mode = strings.ToUpper(opts.metricThresholdMode)
+		alert.MetricThreshold = metric
```
Comment (gssbzn) [inline internal/cli/atlas_alertconfig_create.go]: same, can this be encapsulated?

File: internal/cli/atlas_alertconfig_create.go [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+		notification.EmailEnabled = &opts.notificationEmailEnabled
+
+	case ops:
+		notification.OpsGenieAPIKey = opts.apiKey
+		notification.OpsGenieRegion = opts.notificationRegion
+
+	case pager:
+		notification.ServiceKey = opts.notificationServiceKey
+
+	}
```
Comment (gssbzn) [inline internal/cli/atlas_alertconfig_create.go]: this can be its own function as well

File: internal/cli/atlas_alertconfig_create.go [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+
+	alert.Notifications = []atlas.Notification{*notification}
```
Comment (gssbzn) [inline internal/cli/atlas_alertconfig_create.go]: instead of doing `new(Notification)` you can do `:= Notification{}` and not use a pointer here
also please consider that the following is also valid/ok
```go
notification := atlas.Notification{
TypeName: strings.ToUpper(opts.notificationTypeName),
DelayMin: &opts.notificationDelayMin,
IntervalMin: opts.notificationIntervalMin,
TeamID: opts.notificationTeamID,
Username: opts.notificationUsername,
}
```

**Comment by gssbzn** [comment]
@andreaangiolillo another comment, please remember to update the PR description marking the tasks that you have done, the CLA one is not needed for mdb employees so you can delete that

File: internal/cli/atlas_alertconfig_create.go (hunk 2 above)
Comment (andreaangiolillo) [reply internal/cli/atlas_alertconfig_create.go]: I am also going to rename atlas_alertconfig_list

File: internal/flags/flags.go [hunk 14]
```diff
@@ -15,32 +15,59 @@
...
+	Event                           = "event"                           // EventTypeName flag
+	Enabled                         = "enabled"                         // Enabled flag
+	MatcherFieldName                = "matcherFieldName"                // MatcherFieldName flag
+	MatcherOperator                 = "matcherOperator"                 // MatcherOperator flag
+	MatcherValue                    = "matcherValue"                    // MatcherValue flag
+	MetricThresholdMetricName       = "metricThresholdMetricName"       // MetricThresholdMetricName flag
+	MetricThresholdOperator         = "metricThresholdOperator"         // MetricThresholdOperator flag
+	MetricThresholdThreshold        = "metricThresholdThreshold"        // MetricThresholdThreshold flag
+	MetricThresholdUnits            = "metricThresholdUnits"            // MetricThresholdUnits flag
+	MetricThresholdMode             = "metricThresholdMode"             // MetricThresholdMode flag
```
Comment (gssbzn) [inline internal/flags/flags.go]: - `--metricThresholdMetricName`, double metric, what about just `--metricName` or just `--metric`
- `--metricThresholdOperator`, maybe just `metricOperator`
- `--metricThresholdThreshold `, double threshold, what about `metricThreshold` ?
- `--metricThresholdUnits`, `--metricThresholdMode` to keep things in line maybe also just `metricUnits` and `--metricMode`

Comment (gssbzn) [reply internal/flags/flags.go]: not needed but consider changing the const name to match

File: internal/cli/atlas_alert_config_create.go [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+
+func buildMatcher(opts *atlasAlertConfigCreateOpts, alertConfig *atlas.AlertConfiguration) {
+	if opts.matcherFieldName != "" {
+		match := new(atlas.Matcher)
+		match.FieldName = strings.ToUpper(opts.matcherFieldName)
+		match.Operator = strings.ToUpper(opts.matcherOperator)
+		match.Value = strings.ToUpper(opts.matcherValue)
+		alertConfig.Matchers = []atlas.Matcher{*match}
+	}
+}
```
Comment (gssbzn) [inline internal/cli/atlas_alert_config_create.go]: there are multiple approaches for this, I don't think one is particularly better than the other
If you want to keep the approach where you modify an existing struct please reduce the area of your functions, don't pass the whole `atlas.AlertConfiguration`, pass the appropriate struct for the method name, be it a  `atlas.MetricThreshold` or `atlas.Matcher`, also please name this variable `out` to indicate to users of your function that that struct is being modified

Another approach you could have done:
```go
func (opts *atlasAlertConfigCreateOpts) newMatcher() *atlas.Matcher {
  return &atlas.Matcher{
    FieldName: strings.ToUpper(opts.matcherFieldName),
    Operator: strings.ToUpper(opts.matcherOperator),
    Value: strings.ToUpper(opts.matcherValue),
  }
}
```
And build the array in `buildAlertConfiguration` after doing the check for `opts.matcherFieldName != ""` in that method, same with `buildMetricThreshold`, and `buildNotification`, also the use of `new*` for methods that create a new reference to a struct is more common on go than `build*`

File: internal/cli/atlas_alert_config.go [hunk 16]
```diff
@@ -25,6 +25,8 @@ func AtlasAlertConfigBuilder() *cobra.Command {
 		Short:   "Manage Atlas alert configurations for your project.",
 		Long:    "The alertConfigs command provides access to your alerts configurations. You can create, edit, and delete alert configurations.",
 	}
+
+	cmd.AddCommand(AtlasAlertConfigCreateBuilder())
```
Comment (gssbzn) [inline internal/cli/atlas_alert_config.go]: this is not your fault but mine as I designed the naming, but after checking out your branch and testing it I'm not convince about `mcli atlas alert-configs cmd` structure, I think we may change this to be `mcli atlas alerts configs cmd` and let it leave next to the `mcli atlas alerts cmd`

File: internal/usage/usage.go (hunk 5 above)
Comment (gssbzn) [reply internal/usage/usage.go]: @andreaangiolillo all these changes to spacing here breaks some help commands
Screenshot 2020-02-24 at 11 07 11

File: internal/cli/atlas_alert_config.go [hunk 17]
```diff
@@ -20,11 +20,13 @@ import (
 
 func AtlasAlertConfigBuilder() *cobra.Command {
 	cmd := &cobra.Command{
-		Use:     "alert-config",
-		Aliases: []string{"alertConfig", "alertConfigs", "alert-configs"},
+		Use:     "alerts configs",
```
Comment (gssbzn) [inline internal/cli/atlas_alert_config.go]: if you wanted to do this now, please create a file `internal/cli/atlas_alerts.go` and define `alerts` there and add the config command to that one
Also only add `alert` as an alias

Comment (gssbzn) [reply internal/cli/atlas_alert_config.go]: once you create that file this should only be `configs` and only add `config` as an alias

File: internal/cli/atlas_alert_config_create.go [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+	if err != nil {
+		return err
+	}
+
+	return json.PrettyPrint(result)
+}
+
+func (opts *atlasAlertConfigCreateOpts) buildAlertConfiguration() *atlas.AlertConfiguration {
+
+	out := new(atlas.AlertConfiguration)
```
Comment (gssbzn) [inline internal/cli/atlas_alert_config_create.go]: FYI `out` is only needed when is a parameter, see https://github.com/mongodb/mcli/blob/0f7be0649ffe41a9d782b300fa86f7f0fa7b8390/internal/file/file.go#L17

If is an internal variable it doesn't matter much, no need to do anything here, just letting you know

File: internal/usage/usage.go [hunk 19]
```diff
@@ -34,14 +34,39 @@ const (
...
+	NotificationFlowName            = "Flowdock flow name in lower-case letters"
+	NotificationIntervalMin         = "Number of minutes to wait between successive notifications for unacknowledged alerts that are not resolved"
+	NotificationMobileNumber        = "Mobile number to which alert notifications are sent"
+	NotificationOrgName             = "Flowdock organization name in lower-case letters"
+	NotificationServiceKey          = "PagerDuty service key"
+	NotificationSmsEnabled          = "Flag indicating if text message notifications should be sent"
+	NotificationTeamID              = "Unique identifier of a team"
+	NotificationType                = "Type of alert notification"
+	NotificationUsername            = "Name of the Atlas user to which to send notifications"
+	NotificationVictorOpsRoutingKey = "VictorOps routing key"
```
Comment (gssbzn) [inline internal/usage/usage.go]: Please add a period(.) at the end of the help text

File: internal/cli/atlas_alert_config_create_test.go [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+	"github.com/mongodb/mcli/internal/mocks"
+)
+
+func TestAtlasAlertConfigCreate_Run(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mockStore := mocks.NewMockAlertConfigurationCreator(ctrl)
+
+	defer ctrl.Finish()
+
+	t.Run("flags run", func(t *testing.T) {
```
Comment (gssbzn) [inline internal/cli/atlas_alert_config_create_test.go]: Since this won't run from a file I think is safe to move the test out of this closure

File: internal/cli/atlas_alert_config.go [hunk 21]
```diff
@@ -18,10 +18,10 @@ import (
 	"github.com/spf13/cobra"
 )
 
-func AtlasAlertConfigBuilder() *cobra.Command {
+func AtlasAlertConfigsBuilder() *cobra.Command {
 	cmd := &cobra.Command{
-		Use:     "alerts configs",
-		Aliases: []string{"alertConfig", "alertConfigs", "alert-configs", "alert-config"},
+		Use:     "configs",
+		Aliases: []string{"alertConfig", "alertConfigs", "alert-configs", "alert-config", "config"},
```
Comment (gssbzn) [inline internal/cli/atlas_alert_config.go]: please remove `"alertConfig", "alertConfigs", "alert-configs", "alert-config"` other wise the command would be `alerts alert-config cmd`

File: internal/usage/usage.go [hunk 22]
```diff
@@ -15,31 +15,56 @@
...
+	Password                        = "User’s password."
+	Roles                           = "User's roles and the databases or collections on which the roles apply."
+	Comment                         = "Optional description of the whitelist entry."
+	Force                           = "Don't ask for confirmation."
+	Email                           = "User’s email address."
+	FirstName                       = "User’s first name."
+	LastName                        = "User’s last name."
+	Filename                        = "Filename to use"
+	WhitelistIps                    = "IP addresses to add to the new user’s whitelist."
+	WhitelistType                   = `Type of whitelist entry. Valid values: cidrBlock|ipAddress`
```
Comment (gssbzn) [inline internal/usage/usage.go]: @andreaangiolillo just caught this, this used to be multi line helps, please revert the change if move them with the other multiline texts

File: internal/flags/flags.go [hunk 23]
```diff
@@ -15,32 +15,59 @@
...
+	MatcherFieldName                = "matcherFieldName"                // MatcherFieldName flag
+	MatcherOperator                 = "matcherOperator"                 // MatcherOperator flag
+	MatcherValue                    = "matcherValue"                    // MatcherValue flag
+	MetricName                      = "metricName"                      // MetricName flag
+	MetricOperator                  = "metricOperator"                  // MetricOperator flag
+	MetricThreshold                 = "metricThreshold"                 // MetricThreshold flag
+	MetricUnits                     = "metricUnits"                     // MetricUnits flag
+	MetricMode                      = "metricMode"                      // MetricMode flag
+	NotificationToken               = "notificationToken"               // NotificationToken flag
+	NotificationChannelName         = "notificationsChannelName"        // NotificationChannelName flag
```
Comment (gssbzn) [inline internal/flags/flags.go]: extra `s`, `notificationsChannelName` -> `notificationChannelName`

File: internal/cli/atlas_alerts.go [hunk 24]
```diff
@@ -0,0 +1,18 @@
...
+
+import (
+	"github.com/spf13/cobra"
+)
+
+func AtlasAlertsBuilder() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "alerts",
+		Aliases: []string{"alert"},
+		Short:   "Manage Atlas alert for your project.",
```
Comment (gssbzn) [inline internal/cli/atlas_alerts.go]: [nit] `Manage Atlas alerts for your project.`

File: internal/cli/atlas_alerts.go [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+		Long:    "The alertConfigs command provides access to your alerts configurations. You can create, edit, and delete alert configurations.",
```
Comment (gssbzn) [inline internal/cli/atlas_alerts.go]: [nit] delete this line as it does not apply to this level

File: internal/cli/atlas_alert_config_create.go [hunk 26]
```diff
@@ -0,0 +1,236 @@
...
+// [--notificationType type --notificationDelayMin min --notificationEmailEnabled --notificationSmsEnabled --notificationUsername username --notificationTeamID id
+// --notificationEmailAddress email --notificationMobileNumber number --notificationChannelName channel --notificationApiToken --notificationRegion region] [--projectId projectId]
+func AtlasAlertConfigCreateBuilder() *cobra.Command {
+	opts := &atlasAlertConfigCreateOpts{
+		globalOpts: newGlobalOpts(),
+	}
+	cmd := &cobra.Command{
+		Use:     "create",
+		Short:   "Create an Atlas alert configuration for a project.",
+		Aliases: []string{"cr", "Create"},
```
Comment (gssbzn) [inline internal/cli/atlas_alert_config_create.go]: Is `cr` a common alias for cli? I don't think I've seen it before but happy to be proven wrong, also we don't offer alias with caps for other commands so we can safety remove `Create`

File: internal/cli/atlas_alert_config.go [hunk 27]
```diff
@@ -18,13 +18,15 @@ import (
...
 
-func AtlasAlertConfigBuilder() *cobra.Command {
+func AtlasAlertConfigsBuilder() *cobra.Command {
 	cmd := &cobra.Command{
-		Use:     "alert-config",
-		Aliases: []string{"alertConfig", "alertConfigs", "alert-configs"},
+		Use:     "configs",
+		Aliases: []string{"config"},
 		Short:   "Manage Atlas alert configurations for your project.",
 		Long:    "The alertConfigs command provides access to your alerts configurations. You can create, edit, and delete alert configurations.",
```
Comment (gssbzn) [inline internal/cli/atlas_alert_config.go]: this has changed to be just `configs` now
